### PR TITLE
feat: Bump hive_metastore to use pure rust thrift impl `volo`

### PR DESCRIPTION
**Pull request by Xuanwo**
Close https://github.com/apache/iceberg-rust/pull/155

### CODE REVIEW

File: crates/catalog/hms/Cargo.toml [hunk 1]
```diff
@@ -28,13 +28,9 @@ license = { workspace = true }
...
+hive_metastore = "0.0.2"
 iceberg = { workspace = true }
-# the thrift upstream suffered from no regular rust release.
-#
-# [test-rs](https://github.com/tent-rs) is an organization that helps resolves this
-# issue. And [tent-thrift](https://github.com/tent-rs/thrift) is a fork of the thrift
-# crate, built from the thrift upstream with only version bumped.
-thrift = { package = "tent-thrift", version = "0.18.1" }
 typed-builder = { workspace = true }
+volo-thrift = "0.9.2"
```
Comment (liurenjie1024) [inline crates/catalog/hms/Cargo.toml:36]: How about moving this and `hive_metastore` to workspace?

File: crates/catalog/hms/src/catalog.rs [hunk 2]
```diff
@@ -71,19 +54,29 @@ impl Debug for HmsCatalog {
...
+                    ErrorKind::Unexpected,
+                    format!("invalid address: {}", config.address),
+                )
+            })?;
+
+        let client = ThriftHiveMetastoreClientBuilder::new("hms")
+            .address(address)
+            // Framed thrift rpc is not enabled by default in HMS, we use
+            // buffered instead.
+            .make_codec(volo_thrift::codec::default::DefaultMakeCodec::buffered())
```
Comment (liurenjie1024) [inline crates/catalog/hms/src/catalog.rs:74]: Maybe it's better to make this a config?

File: crates/catalog/hms/src/utils.rs [hunk 3]
```diff
@@ -15,13 +15,28 @@
...
+{
     Error::new(
         ErrorKind::Unexpected,
         "operation failed for hitting thrift error".to_string(),
     )
+    .with_source(anyhow!("thrift error: {:?}", error))
+}
+
+/// Format an io error into iceberg error.
+pub fn from_io_error(error: io::Error) -> Error {
```
Comment (liurenjie1024) [inline crates/catalog/hms/src/utils.rs:36]: `io::Error` is a common error, how about moving it to core crate?

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM, thanks!

**Comment by liurenjie1024** [comment]
cc @Xuanwo Any updates?

File: crates/catalog/hms/src/utils.rs (hunk 3 above)
Comment (Xuanwo) [reply crates/catalog/hms/src/utils.rs:36]: I'm not sure if moving to core and making it a public API is a good idea. What if we address this once we have the same `from_io_error` in core?

File: crates/catalog/hms/src/catalog.rs (hunk 2 above)
Comment (Xuanwo) [reply crates/catalog/hms/src/catalog.rs:74]: Tracked at https://github.com/apache/iceberg-rust/issues/188

**Comment by Xuanwo** [comment]
cc @liurenjie1024, please review again. Sorry for the late.

**Review by liurenjie1024 — APPROVED** [review]
LGTM, let's move!

**Review by Fokko — APPROVED** [review]
Thanks for the PR @Xuanwo and @liurenjie1024 for the review 🙌

**Pull request merged**